starlette/routing.py

```diff
@@ -342,14 +350,21 @@ def __init__(
             self.app: ASGIApp = app
         else:
             self.app = Router(routes=routes)
+        self._user_app = self.app
         self.name = name
         self.path_regex, self.path_format, self.param_convertors = compile_path(
             self.path + "/{path:path}"
         )
 
+        if middleware is not None:
+            for cls, options in reversed(middleware):
+                self.app = cls(app=self.app, **options)
+
     @property
     def routes(self) -> typing.List[BaseRoute]:
-        return getattr(self.app, "routes", [])
+        # we dynamically grab the routes so that if this is a Starlette router
+        # it can have routes added to it after it is mounted
+        return getattr(self._user_app, "routes", [])
```

Comment:
@tomchristie how would you feel about just grabbing a reference to `.routes` in `__init__` instead of doing a getattr here? I suppose that would break if the mounted router replaces its `.routes` with a new list, but otherwise it should be fine and it would avoid having to add this `_user_app` attribute.

Reply:
I think I like it more. I went ahead and implemented it that way: https://github.com/encode/starlette/pull/1286/commits/e9ff90328df6285b2ebf77ca5abf2c053d42add6

If you prefer it as it was doing a `getattr` in `self.routes`, I'm happy to revert it